File: Java/DL_SDK_Sample_App/app/src/main/java/com/example/dl_sdk_sample_app/BluetoothSilentPairingFragment.java
package com.example.dl_sdk_sample_app;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.*;
import androidx.fragment.app.Fragment;
import com.datalogic.device.ErrorManager;
import com.datalogic.device.configuration.*;
import com.example.dl_sdk_sample_app.databinding.FragmentBluetoothSilentPairingBinding;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BluetoothSilentPairingFragment extends Fragment {

    private FragmentBluetoothSilentPairingBinding binding;
    private ConfigurationManager configurationManager;
    private final String[] btPropertyNames = {
            "BT_DISCOVERABILITY",
            "BT_PAIRING_POLICY",
            "BT_SILENT_PAIRING_TRUSTED_ENABLE",
            "BT_SILENT_PAIRING_WHITELISTING_ENABLE",
            "BT_SILENT_PAIRING_WHITELISTING"
    };
    private final Map<Property<?>, View> propertyViews = new HashMap<>();

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState){
        binding = FragmentBluetoothSilentPairingBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState){
        super.onViewCreated(view, savedInstanceState);
        ErrorManager.enableExceptions(true);
        try {
            configurationManager = new ConfigurationManager(requireContext());
        } catch (Exception e){
            e.printStackTrace();
            Toast.makeText(getContext(), "Failed to initialize ConfigurationManager", Toast.LENGTH_SHORT).show();
            return;
        }
        setupPropertyInputs();
        binding.buttonApplyBtSettings.setOnClickListener(v -> applyBluetoothSettings());
    }

    private void setupPropertyInputs(){
        List<Property<?>> btProperties = new ArrayList<>();
        for(String name : btPropertyNames){
            try {
                Property<?> prop = configurationManager.getPropertyByName(name);
                if(prop == null){
                    Log.e("BluetoothSilentPairingFragment", "Property '" + name + "' is null. Check if the property name is correct.");
                    Toast.makeText(getContext(), "Property '" + name + "' not found.", Toast.LENGTH_SHORT).show();
                } else {
                    Log.d("BluetoothSilentPairingFragment", "Property '" + name + "' retrieved successfully.");
                    btProperties.add(prop);
                }
            } catch (Exception e){
                Log.e("BluetoothSilentPairingFragment", "Error retrieving property '" + name + "': " + e.getMessage());
            }
        }
        for(Property<?> property : btProperties){
            if(property.isSupported() && !property.isReadOnly()){
                View inputView = createInputForProperty(property);
                if(inputView != null){
                    binding.linearLayoutBtProperties.addView(inputView);
                    propertyViews.put(property, inputView);
                }
            }
        }
    }

    private View createInputForProperty(Property<?> property){
        Context context = requireContext();
        String propertyName = property.getName();
        LinearLayout container = new LinearLayout(context);
        container.setOrientation(LinearLayout.HORIZONTAL);
        container.setLayoutParams(new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
        ));
        container.setPadding(0, 8, 0, 8);

        TextView label = new TextView(context);
        label.setText(propertyName);
        label.setLayoutParams(new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f));

        if(property instanceof BooleanProperty){
            Switch switchView = new Switch(context);
            switchView.setChecked(((BooleanProperty) property).get());
            container.addView(label);
            container.addView(switchView);
            return container;
        } else if(property instanceof EnumProperty<?>){
            Spinner spinner = new Spinner(context);
            Enum<?>[] enumConstants = ((EnumProperty<?>) property).getEnumConstants();
            if(enumConstants != null){
                String[] enumNames = new String[enumConstants.length];
                for(int i=0;i<enumConstants.length;i++) enumNames[i] = enumConstants[i].toString();
                ArrayAdapter<String> adapter = new ArrayAdapter<>(context, android.R.layout.simple_spinner_item, enumNames);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                spinner.setAdapter(adapter);
                Object currentValue = ((EnumProperty<?>) property).get();
                int index = java.util.Arrays.asList(enumConstants).indexOf(currentValue);
                if(index >=0 ) spinner.setSelection(index);
            }
            container.addView(label);
            container.addView(spinner);
            return container;
        } else if(property instanceof TextProperty){
            EditText editText = new EditText(context);
            editText.setInputType(android.text.InputType.TYPE_CLASS_TEXT);
            editText.setText(((TextProperty) property).get());
            container.addView(label);
            container.addView(editText);
            return container;
        } else {
            Toast.makeText(getContext(), "Unsupported property type: " + property.getType(), Toast.LENGTH_SHORT).show();
            return null;
        }
    }

    @SuppressWarnings("unchecked")
    private void applyBluetoothSettings(){
        try {
            for(Map.Entry<Property<?>, View> entry : propertyViews.entrySet()){
                Property<?> property = entry.getKey();
                View view = entry.getValue();
                if(property instanceof BooleanProperty){
                    boolean value = ((Switch)((LinearLayout)view).getChildAt(1)).isChecked();
                    ((BooleanProperty) property).set(value);
                } else if(property instanceof EnumProperty){
                    Spinner spinner = (Spinner)((LinearLayout)view).getChildAt(1);
                    int selectedPosition = spinner.getSelectedItemPosition();
                    Enum<?>[] enums = ((EnumProperty<?>) property).getEnumConstants();
                    if(enums != null && selectedPosition < enums.length){
                        Enum<?> enumValue = enums[selectedPosition];
                        try {
                            Method setMethod = property.getClass().getMethod("set", enumValue.getClass());
                            setMethod.invoke(property, enumValue);
                        } catch (Exception e){
                            e.printStackTrace();
                            Toast.makeText(getContext(), "Set method not found for property: " + property.getName(), Toast.LENGTH_SHORT).show();
                        }
                    }
                } else if(property instanceof TextProperty){
                    String input = ((EditText)((LinearLayout)view).getChildAt(1)).getText().toString();
                    ((TextProperty) property).set(input);
                }
            }
            configurationManager.commit();
            Toast.makeText(getContext(), "Bluetooth settings applied successfully", Toast.LENGTH_SHORT).show();
        } catch (ConfigException e){
            e.printStackTrace();
            Toast.makeText(getContext(), "Failed to apply Bluetooth settings: " + e.getMessage(), Toast.LENGTH_LONG).show();
        } catch (Exception e){
            e.printStackTrace();
            Toast.makeText(getContext(), "An unexpected error occurred while applying settings.", Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void onDestroyView(){
        super.onDestroyView();
        binding = null;
    }
}

File: Kotlin/DL_SDK_Sample_App/app/src/main/java/com/example/dl_sdk_sample_app/BluetoothSilentPairingFragment.kt
package com.example.dl_sdk_sample_app

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.datalogic.device.ErrorManager
import com.datalogic.device.configuration.*
import com.example.dl_sdk_sample_app.databinding.FragmentBluetoothSilentPairingBinding

class BluetoothSilentPairingFragment : Fragment() {

    private var _binding: FragmentBluetoothSilentPairingBinding? = null
    private val binding get() = _binding!!

    private lateinit var configurationManager: ConfigurationManager

    // List of Bluetooth-related property names
    private val btPropertyNames = listOf(
        "BT_DISCOVERABILITY",
        "BT_PAIRING_POLICY",
        "BT_SILENT_PAIRING_TRUSTED_ENABLE",
        "BT_SILENT_PAIRING_WHITELISTING_ENABLE",
        "BT_SILENT_PAIRING_WHITELISTING"
    )

    // Map to store Property objects and their corresponding Views
    private val propertyViews = mutableMapOf<Property<*>, View>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        _binding = FragmentBluetoothSilentPairingBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        // Enable exception handling for Datalogic SDK
        ErrorManager.enableExceptions(true)

        // Initialize ConfigurationManager with Context
        try {
            configurationManager = ConfigurationManager(requireContext())
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(context, "Failed to initialize ConfigurationManager", Toast.LENGTH_SHORT).show()
            return
        }

        // Fetch and setup property inputs
        setupPropertyInputs()

        // Set listener for Apply button
        binding.buttonApplyBtSettings.setOnClickListener {
            applyBluetoothSettings()
        }
    }

    /**
     * Fetches Bluetooth-related properties and sets up input views.
     */
    private fun setupPropertyInputs() {
        val btProperties = btPropertyNames.mapNotNull { propertyName ->
            try {
                configurationManager.getPropertyByName(propertyName)
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }

        for (property in btProperties) {
            if (property.isSupported && !property.isReadOnly) {
                val inputView = createInputForProperty(property)
                if (inputView != null) {
                    binding.linearLayoutBtProperties.addView(inputView)
                    propertyViews[property] = inputView
                }
            }
        }
    }

    /**
     * Creates an appropriate input view based on the property's type.
     */
    private fun createInputForProperty(property: Property<*>): View? {
        val context = requireContext()
        val propertyName = property.name

        val container = LinearLayout(context).apply {
            orientation = LinearLayout.HORIZONTAL
            layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )
            setPadding(0, 8, 0, 8)
        }

        val label = TextView(context).apply {
            text = propertyName
            layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f)
        }

        when (property) {
            is BooleanProperty -> {
                val switchView = Switch(context).apply {
                    isChecked = property.get()
                }
                container.addView(label)
                container.addView(switchView)
                return container
            }
            is EnumProperty<*> -> {
                val spinner = Spinner(context)
                val enumValues = property.enumConstants

                if (enumValues != null) {
                    val enumNames = enumValues.map { it.toString() }
                    val adapter = ArrayAdapter(
                        context,
                        android.R.layout.simple_spinner_item,
                        enumNames
                    )
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                    spinner.adapter = adapter

                    // Set current selection
                    val currentValue = property.get()
                    val index = enumValues.indexOf(currentValue)
                    if (index >= 0) {
                        spinner.setSelection(index)
                    }
                }

                container.addView(label)
                container.addView(spinner)
                return container
            }
            is TextProperty -> {
                val editText = EditText(context).apply {
                    inputType = android.text.InputType.TYPE_CLASS_TEXT
                    setText(property.get())
                }
                container.addView(label)
                container.addView(editText)
                return container
            }
            else -> {
                Toast.makeText(context, "Unsupported property type: ${property.type}", Toast.LENGTH_SHORT).show()
                return null
            }
        }
    }

    /**
     * Applies the Bluetooth settings based on user input.
     */
    private fun applyBluetoothSettings() {
        try {
            for ((property, view) in propertyViews) {
                when (property) {
                    is BooleanProperty -> {
                        val switchView = (view as LinearLayout).getChildAt(1) as Switch
                        property.set(switchView.isChecked)
                    }
                    is EnumProperty<*> -> {
                        val spinner = (view as LinearLayout).getChildAt(1) as Spinner
                        val selectedPosition = spinner.selectedItemPosition
                        val enumValues = property.enumConstants

                        if (enumValues != null && selectedPosition in enumValues.indices) {
                            val enumValue = enumValues[selectedPosition]
                            // Use reflection to invoke the set method
                            try {
                                val setMethod = property.javaClass.getMethod("set", enumValue.javaClass)
                                setMethod.invoke(property, enumValue)
                            } catch (e: NoSuchMethodException) {
                                e.printStackTrace()
                                Toast.makeText(context, "Set method not found for property: ${property.name}", Toast.LENGTH_SHORT).show()
                            } catch (e: Exception) {
                                e.printStackTrace()
                                Toast.makeText(context, "Failed to set property: ${property.name}", Toast.LENGTH_SHORT).show()
                            }
                        }
                    }
                    is TextProperty -> {
                        val editText = (view as LinearLayout).getChildAt(1) as EditText
                        val input = editText.text.toString()
                        property.set(input)
                    }
                }
            }

            // Commit the changes
            configurationManager.commit()
            Toast.makeText(context, "Bluetooth settings applied successfully", Toast.LENGTH_SHORT).show()
        } catch (e: ConfigException) {
            e.printStackTrace()
            Toast.makeText(context, "Failed to apply Bluetooth settings: ${e.message}", Toast.LENGTH_LONG).show()
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(context, "An unexpected error occurred while applying settings.", Toast.LENGTH_LONG).show()
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
